### PR TITLE
Yf use mapped file dictionary v2

### PR DESCRIPTION
**Pull request by yfarjoun**
- Changing MergeBamAlignment so that it merges the dictionaries of the aligned bam and the reference instead of taking the reference dict file. This is since bwa-mem (latest versions) add a AH:* to the sequence records of "Alternate Haplotypes" and this is lost in translation otherwise.

### CODE REVIEW

**Comment by yfarjoun** [comment]
@tfenne given that you reviewed the first attempt at this PR, would you be available to review this version? as a reminder, the previous version is here: #627

**Comment by yfarjoun** [comment]
@tfenne got time to look at this?

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 1]
```diff
@@ -284,36 +254,29 @@ public AbstractAlignmentMerger(final File unmappedBamFile, final File targetBamF
...
-        }
 
         this.clipAdapters = clipAdapters;
         this.bisulfiteSequence = bisulfiteSequence;
         this.alignedReadsOnly = alignedReadsOnly;
 
         this.header = new SAMFileHeader();
-        this.sortOrder = sortOrder != null ? sortOrder : SortOrder.coordinate;
-        header.setSortOrder(SortOrder.coordinate);
+        this.sortOrder = sortOrder != null ? sortOrder : SAMFileHeader.SortOrder.coordinate;
```
Comment (tfenne) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: Why the unnecessary changes?

File: src/main/java/picard/sam/SamAlignmentMerger.java [hunk 2]
```diff
@@ -45,6 +31,7 @@
     private final int minUnclippedBases;
     private boolean forceSort = false;
     private final OverclippedReadFilter contaminationFilter;
+    private final List<String> matchingDictionaryTags;
```
Comment (tfenne) [inline src/main/java/picard/sam/SamAlignmentMerger.java]: Are these the ones required to match?  Or the ones that did match?

File: src/main/java/picard/sam/SamAlignmentMerger.java [hunk 3]
```diff
@@ -145,7 +136,8 @@ public SamAlignmentMerger(final File unmappedBamFile, final File targetBamFile,
                               final boolean addMateCigar,
                               final boolean unmapContaminantReads,
                               final int minUnclippedBases,
-                              final UnmappingReadStrategy unmappingReadStrategy) {
+                              final UnmappingReadStrategy unmappingReadStrategy,
+                              final List<String> matchingDictionaryTags) {
```
Comment (tfenne) [inline src/main/java/picard/sam/SamAlignmentMerger.java]: -> requiredMatchingDictionaryTags

File: src/main/java/picard/sam/SamAlignmentMerger.java [hunk 4]
```diff
@@ -194,6 +186,27 @@ public void mergeAlignment(final File referenceFasta) {
...
+                    .forEach(finalDict::assertSameDictionary);
+        } else {
+            final SeparateEndAlignmentIterator mergingIterator = new SeparateEndAlignmentIterator(this.read1AlignedSamFile, this.read2AlignedSamFile, referenceFasta);
+            finalDict = mergingIterator.getHeader().getSequenceDictionary();
+        }
+
+        SAMSequenceDictionary referenceDict = SAMSequenceDictionaryExtractor.extractDictionary(referenceFasta);
+        if (referenceDict == null) {
+            throw new PicardException("No sequence dictionary found for " + referenceFasta.getAbsolutePath() +
+                    ".  Use CreateSequenceDictionary.jar to create a sequence dictionary.");
```
Comment (tfenne) [inline src/main/java/picard/sam/SamAlignmentMerger.java]: This message is horribly out of date ;)  `Use Picard's CreateSequenceDictionary`

**Comment by yfarjoun** [comment]
Sorry about the cosmetic changes. I need to find a good way to separate the changes like that...

I'll fix the conflict (again) and merge.

**Comment by yfarjoun** [comment]
rebase was needed. waiting for tests and then merging.
